src/puppetlabs/puppetdb/scf/storage.clj

```diff
@@ -1364,14 +1347,31 @@
                               AND (deactivated IS NULL OR deactivated < ?)"
                          [sql-timestamp certname sql-timestamp])))))
 
-(pls/defn-validated expire-node!
-  "Expire the given host, recording the current time. If the node is
-  currently expired, no change is made."
-  [certname :- String & [timestamp :- pls/Timestamp]]
-  (let [timestamp (to-timestamp (or timestamp (now)))]
-    (jdbc/do-prepared "UPDATE certnames SET expired = ?
-                         WHERE certname=? AND expired IS NULL"
-                      [timestamp certname])))
+(pls/defn-validated expire-stale-nodes [horizon :- Period]
+  "Expires nodes with no activity within the provided horizon (prior
+  to now) and returns a collection of the affected certnames."
+  (let [stale-start-ts (to-timestamp (ago horizon))
+        expired-ts (to-timestamp (now))]
+    (map :certname
+         (jdbc/query-to-vec
+          (str
+           "update certnames set expired = ?"
+           "  where id in"
+           "    (select c.id from certnames c"
+           "         left outer join catalogs cats on c.certname=cats.certname"
+           "         left outer join factsets fs on c.certname=fs.certname"
+           "         left outer join reports r on c.certname=r.certname"
```

Comment:
It might be more efficient to join on certname.latest_report_id which will only return one row from reports instead of all of the reports.  

It's unfortunate that we need to join all the tables to determine inactivity.